### PR TITLE
OCPVE-168 kdump logs for failed realtime tasks

### PR DESCRIPTION
**Pull request by jeff-roche**
This change is to configure and collect logs from kdump in the event of a kernel panic. This will help debug realtime test failures that occur due to any issues with the realtime kernel

### CODE REVIEW

**Comment by jeff-roche** [comment]
holding until the tests are passingi
/hold

File: ci-operator/step-registry/optional-operators/kdump-configure-logs/optional-operators-kdump-configure-logs-commands.sh [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+  kernelArguments:
+    - crashkernel=256M
+EOF
+
+echo "waiting for mcp/${node_role} condition=Updating timeout=5m"
+oc wait mcp/${node_role} --for condition=Updating --timeout=5m
+
+# When applying this configuration to the master node for single node
+# we need to wait for the master node to restart before we attempt to call out to the API Server
+if [[ "$node_role" == "master" ]]; then
```
Comment (eggfoobar) [inline ci-operator/step-registry/optional-operators/kdump-configure-logs/optional-operators-kdump-configure-logs-commands.sh:63]: Might be best to add a single node topology check here as part of the if statement. I'm not 100% sure if `oc wait` will fail if the API server isn't responding, which will be the case during single node upgrade. So maybe something like this 
```
if node_role == master && single_node_topo; then 
wait 600seconds; 
else 
oc wait mcp/${node_role} --for condition=Updating --timeout=5m 
fi
```
 What do you think?

Comment (jeff-roche) [reply ci-operator/step-registry/optional-operators/kdump-configure-logs/optional-operators-kdump-configure-logs-commands.sh:63]: Doesn't `node_role` == "master" essentially check if it's SNO? It would be "worker" for non-sno setups

Comment (eggfoobar) [reply ci-operator/step-registry/optional-operators/kdump-configure-logs/optional-operators-kdump-configure-logs-commands.sh:63]: Not necessarily, `node_role` is identifying the the label we want to apply this machine config to, but you could have 3 master nodes, so `node_role` would still be `master` but not SNO.

**Comment by jeff-roche** [comment]
None of the rehearsal errors are related to kdump

**Comment by ffromani** [comment]
/hold
I have no context about this change entirely

**Comment by jeff-roche** [comment]
@14rcole This is ready for approval (finally)

**Comment by stbenjam** [comment]
Does this just get written out to the artifacts directory? I think they need to be made not publicly accessible through some other mechanism, can't kdumps contain privileged information which could include cloud creds and other secrets?

File: ci-operator/step-registry/ipi/conf/realtime/kdump-configure-logs/ipi-conf-realtime-kdump-configure-logs-commands.sh [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+    - crashkernel=256M
+EOF
+
+echo "waiting for mcp/${node_role} condition=Updating timeout=5m"
+oc wait mcp/${node_role} --for condition=Updating --timeout=5m
+
+# When applying this configuration to the master node for single node
+# we need to wait for the master node to restart before we attempt to call out to the API Server
+if [[ "$node_role" == "master" && $infra_status == "SingleReplica" ]]; then
+  seconds=600
```
Comment (DennisPeriquet) [inline ci-operator/step-registry/ipi/conf/realtime/kdump-configure-logs/ipi-conf-realtime-kdump-configure-logs-commands.sh:65]: does this have to be 10 minutes? is it possible to do some kind of loop where we wait say 30s and then try to ping to see if the node is done restarting? -- the idea being we wait only long enough with some granularity of about 30s instead of a static 10 min

File: ci-operator/step-registry/ipi/conf/realtime/kdump-gather-logs/ipi-conf-realtime-kdump-gather-logs-commands.sh [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+
+    # Cleanup the debug pod
+    oc delete pod "$debug_pod" -n "default"
+  fi
+}
+
+# Gather all the kdump logs from the identified nodes in parallel
+function gather_kdump_logs {
+  for NODE in ${NODES}; do
+    gather_kdump_logs_from_node "${NODE}" &
```
[the remark this answers is not in the document]
Comment (DennisPeriquet) [reply ci-operator/step-registry/ipi/conf/realtime/kdump-gather-logs/ipi-conf-realtime-kdump-gather-logs-commands.sh:39]: ah, line 77 -- ignore me :)

File: ci-operator/step-registry/ipi/conf/realtime/kdump-gather-logs/ipi-conf-realtime-kdump-gather-logs-commands.sh [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+
+    echo "INFO: Finished packaging the kdump logs"
+  fi
+
+  # Cleanup
+  rm -rf "${output_path}"
+}
+
+node_label="node-role.kubernetes.io/${node_role}"
+NODES="${*:-$(oc get nodes -l ${node_label} -o jsonpath='{.items[?(@.status.nodeInfo.operatingSystem=="linux")].metadata.name}')}"
```
Comment (DennisPeriquet) [inline ci-operator/step-registry/ipi/conf/realtime/kdump-gather-logs/ipi-conf-realtime-kdump-gather-logs-commands.sh:70]: optional request: please add `echo $NODES` in case someone needs to debug this.

**Comment by jeff-roche** [comment]
> Does this just get written out to the artifacts directory? I think they need to be made not publicly accessible through some other mechanism, can't kdumps contain privileged information which could include cloud creds and other secrets?

@stbenjam as far as I know the `novmcoredd` flag should take care of that. This will keep it from dumping the memory which is where the sensitive data would be

File: ci-operator/step-registry/ipi/conf/realtime/kdump-configure-logs/ipi-conf-realtime-kdump-configure-logs-commands.sh (hunk 2 above)
Comment (jeff-roche) [reply ci-operator/step-registry/ipi/conf/realtime/kdump-configure-logs/ipi-conf-realtime-kdump-configure-logs-commands.sh]: We opted to change how we were applying this config, we now output it to a config file and let the install step pick it up

File: ci-operator/config/stackrox/stackrox/stackrox-stackrox-master.yaml
Comment (DennisPeriquet) [inline ci-operator/config/stackrox/stackrox/stackrox-stackrox-master.yaml:1]: why did you need to updated this file?

Comment (DennisPeriquet) [reply ci-operator/config/stackrox/stackrox/stackrox-stackrox-master.yaml:1]: in slack you indicated this was pulled in from a rebase

**Comment by DennisPeriquet** [comment]
there are no dumps -- just subdirs in [kdump.tar](https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/pr-logs/pull/openshift_release/35790/rehearse-35790-pull-ci-openshift-origin-release-4.11-e2e-gcp-ovn-rt-upgrade/1636036834419019776/artifacts/e2e-gcp-ovn-rt-upgrade/ipi-conf-realtime-kdump-gather-logs/artifacts/):

```
$ ls
kdump.tar

$ tar xzvf kdump.tar 
x .
x ci-op-7nyrgb9v-37080-q9649-worker-a-ffn7n_kdump_logs
x ci-op-7nyrgb9v-37080-q9649-worker-b-nd8jk_kdump_logs
x ci-op-7nyrgb9v-37080-q9649-worker-c-9h9qc_kdump_logs

$ tree
.
├── ci-op-7nyrgb9v-37080-q9649-worker-a-ffn7n_kdump_logs
├── ci-op-7nyrgb9v-37080-q9649-worker-b-nd8jk_kdump_logs
├── ci-op-7nyrgb9v-37080-q9649-worker-c-9h9qc_kdump_logs
└── kdump.tar

3 directories, 1 file
```

File: ci-operator/config/stackrox/stackrox/stackrox-stackrox-master.yaml
Comment (jeff-roche) [reply ci-operator/config/stackrox/stackrox/stackrox-stackrox-master.yaml:1]: > there are no dumps -- just subdirs in [kdump.tar](https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/pr-logs/pull/openshift_release/35790/rehearse-35790-pull-ci-openshift-origin-release-4.11-e2e-gcp-ovn-rt-upgrade/1636036834419019776/artifacts/e2e-gcp-ovn-rt-upgrade/ipi-conf-realtime-kdump-gather-logs/artifacts/):
> 
> ```
> $ ls
> kdump.tar
> 
> $ tar xzvf kdump.tar 
> x .
> x ci-op-7nyrgb9v-37080-q9649-worker-a-ffn7n_kdump_logs
> x ci-op-7nyrgb9v-37080-q9649-worker-b-nd8jk_kdump_logs
> x ci-op-7nyrgb9v-37080-q9649-worker-c-9h9qc_kdump_logs
> 
> $ tree
> .
> ├── ci-op-7nyrgb9v-37080-q9649-worker-a-ffn7n_kdump_logs
> ├── ci-op-7nyrgb9v-37080-q9649-worker-b-nd8jk_kdump_logs
> ├── ci-op-7nyrgb9v-37080-q9649-worker-c-9h9qc_kdump_logs
> └── kdump.tar
> 
> 3 directories, 1 file
> ```

@DennisPeriquet the folders will be empty if there was no kernel crash

**Comment by jeff-roche** [comment]
> there are no dumps -- just subdirs in [kdump.tar](https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/pr-logs/pull/openshift_release/35790/rehearse-35790-pull-ci-openshift-origin-release-4.11-e2e-gcp-ovn-rt-upgrade/1636036834419019776/artifacts/e2e-gcp-ovn-rt-upgrade/ipi-conf-realtime-kdump-gather-logs/artifacts/):
> 
> ```
> $ ls
> kdump.tar
> 
> $ tar xzvf kdump.tar 
> x .
> x ci-op-7nyrgb9v-37080-q9649-worker-a-ffn7n_kdump_logs
> x ci-op-7nyrgb9v-37080-q9649-worker-b-nd8jk_kdump_logs
> x ci-op-7nyrgb9v-37080-q9649-worker-c-9h9qc_kdump_logs
> 
> $ tree
> .
> ├── ci-op-7nyrgb9v-37080-q9649-worker-a-ffn7n_kdump_logs
> ├── ci-op-7nyrgb9v-37080-q9649-worker-b-nd8jk_kdump_logs
> ├── ci-op-7nyrgb9v-37080-q9649-worker-c-9h9qc_kdump_logs
> └── kdump.tar
> 
> 3 directories, 1 file
> ```

@DennisPeriquet the directories will be empty if there wasn't a kernel crash
